File: Codeforces/Div 4/Codeforces Round 928 (Div. 4)/G_Vlad_and_Trouble_at_MIT.py
t = int(input())

for _ in range(t):
    n = int(input())
    fa = [-1] + [int(x)-1 for x in input().split()]
    S = input()

    dp1, dp2 = [0] * (n+1), [0] * (n+1)
    
    for i in range(n-1, -1, -1):
        if S[i] == "S":
            dp1[i] = float("inf")
        elif S[i] == "P":
            dp2[i] = float("inf")

        p = fa[i]
        if p != -1:
            dp1[p] += min(dp1[i], dp2[i]+1)
            dp2[p] += min(dp1[i]+1, dp2[i])
        
    print(min(dp1[0], dp2[0]))

File: Codeforces/Div 4/Codeforces Round 928 (Div. 4)/G_Vlad_and_Trouble_at_MIT.cpp
#include <bits/stdc++.h>
using namespace std;
using LL = long long;
const int INF = 0x3f3f3f3f;
const int MOD = 1e9 + 7;
const int N = 100005;

vector<int> g[N];
vector<int> dp1(N), dp2(N);
string S;

void dfs(int u) {
    dp1[u] = 0;
    dp2[u] = 0;
    for (auto v : g[u]) {
        dfs(v);
        dp1[u] += min(dp1[v], dp2[v] + 1);
        dp2[u] += min(dp1[v] + 1, dp2[v]);
    }
    if (S[u] == 'S') dp1[u] = INF;
    else if (S[u] == 'P') dp2[u] = INF;
}

int main() {
    ios::sync_with_stdio(false); cin.tie(nullptr); cout.tie(nullptr);
    int t;
    cin >> t;
    while (t--) {
        int n; cin >> n;
        for (int i = 0; i < n; ++i) g[i].clear();
        int fa;
        for (int i = 1; i < n; ++i) {
            cin >> fa;
            g[fa - 1].push_back(i);
        }
        cin >> S;
        dfs(0);
        cout << min(dp1[0], dp2[0]) << '\n';
    }
    return 0;
}